aosp_diff/preliminary/system/core/02_0002-Enable-virtual-ab-compression-feature.patch

```diff
@@ -0,0 +1,120 @@
+__system_properties_init during the libc init stage of snapuserd process, it will
+go to ContextsSplit.Initialize finally, and in this function, it will try to read
+/system/etc/selinux/plat_property_contexts. However, system partition should be
+prepared by snapuserd which is not started yes, this will halt both init and snapuserd
+processes:
+[   12.367545] init: Relaunched snapuserd with pid: 211
+[   12.382032] init: init second stage started!
+[   16.434536] I/O error: sector 550576: no user-space daemon for dm-user/system_b-user-cow target pid:128 tgid:128
+[   16.436606] I/O error: sector 550568: no user-space daemon for dm-user/system_b-user-cow target pid:128 tgid:128
+[   16.439396] I/O error: sector 550560: no user-space daemon for dm-user/system_b-user-cow target pid:128 tgid:128
+[   16.441467] I/O error: sector 550552: no user-space daemon for dm-user/system_b-user-cow target pid:128 tgid:128
+[   20.466540] I/O error: sector 5792: no user-space daemon for dm-user/system_b-user-cow target pid:128 tgid:128
+[   20.468463] device-mapper: verity-fec: 252:4: FEC 2949120: read failed (720): -5 pid:125 tgid:125
+
+It because if init second stage is started before snapuserd, /dev/__properties__
+will be created by init. Then in the function of __system_properties_init during
+the libc init stage of snapuserd process, it will go to ContextsSplit.Initialize
+finally, and in this function, it will try to read /system/etc/selinux/plat_property_contexts.
+However, system partition should be prepared by snapuserd which is not started yes,
+this will halt both init and snapuserd processes
+
+Already reported a bug to Google, and this patch also will be removed
+once this issue is resovled.
+
+Signed-off-by: ji, zhenlong z <zhenlong.z.ji@intel.com>
+---
+ fs_mgr/libsnapshot/Android.bp           |  9 +++++++++
+ fs_mgr/libsnapshot/snapuserd_daemon.cpp |  2 ++
+ init/snapuserd_transition.cpp           | 13 +++++++++++++
+ 3 files changed, 24 insertions(+)
+
+diff --git a/fs_mgr/libsnapshot/Android.bp b/fs_mgr/libsnapshot/Android.bp
+index 6a764e4fa..22dbc60b3 100644
+--- a/fs_mgr/libsnapshot/Android.bp
++++ b/fs_mgr/libsnapshot/Android.bp
+@@ -455,6 +455,15 @@ cc_binary {
+     recovery_available: true,
+ }
+ 
++cc_binary {
++    name: "snapuserd_ramdisk",
++    stem: "snapuserd",
++    defaults: ["snapuserd_defaults"],
++    static_executable: true,
++    system_shared_libs: [],
++    ramdisk: true,
++}
++
+ cc_test {
+     name: "cow_api_test",
+     defaults: [
+diff --git a/fs_mgr/libsnapshot/snapuserd_daemon.cpp b/fs_mgr/libsnapshot/snapuserd_daemon.cpp
+index 7fa01b78f..ddfd1ed1c 100644
+--- a/fs_mgr/libsnapshot/snapuserd_daemon.cpp
++++ b/fs_mgr/libsnapshot/snapuserd_daemon.cpp
+@@ -122,6 +122,8 @@ int main(int argc, char** argv) {
+ 
+     android::snapshot::Daemon& daemon = android::snapshot::Daemon::Instance();
+ 
++    LOG(INFO) << "Snapuserd daemon start.";
++    close(open("/dev/.snapuserd", O_WRONLY | O_CREAT | O_CLOEXEC, 0000));
```

Comment:
if open returns a invalid fd, then close with this fd, is there any issue?

Reply:
EBADF will be returned. Actually we reference the code here:
https://cactus-source.iind.intel.com/source/xref/CELADON-MASTER-R2/system/core/init/init.cpp#785